openhcl/virt_mshv_vtl/src/processor/tdx/tlb_flush.rs

```diff
@@ -172,16 +185,95 @@ impl UhProcessor<'_, TdxBacked> {
     }
 }
+
+#[derive(Debug, Inspect)]
+pub(super) struct AtomicTlbRingBuffer {
+    /// The contents of the buffer.
+    #[inspect(hex, with = "|x| inspect::iter_by_index(x.iter())")]
+    buffer: Box<[AtomicU64; FLUSH_GVA_LIST_SIZE]>,
+    /// The number of GVAs that have been added over the lifetime of the VM.
+    gva_list_count: AtomicUsize,
+    /// The number of GVAs that have started being added to the list over the
+    /// lifetime of the VM.
+    in_progress_count: AtomicUsize,
+    /// A guard to ensure that only one thread is writing to the list at a time.
+    write_lock: Mutex<()>,
+}
+
+pub(super) struct AtomicTlbRingBufferWriteGuard<'a> {
+    buf: &'a AtomicTlbRingBuffer,
+    _write_lock: MutexGuard<'a, ()>,
+}
+
+impl AtomicTlbRingBuffer {
+    fn new() -> Self {
+        Self {
+            buffer: Box::new(std::array::from_fn(|_| AtomicU64::new(0))),
+            gva_list_count: AtomicUsize::new(0),
+            in_progress_count: AtomicUsize::new(0),
+            write_lock: Mutex::new(()),
+        }
+    }
+
+    fn count(&self) -> usize {
+        self.gva_list_count.load(Ordering::Relaxed)
+    }
+
+    pub fn write(&self) -> AtomicTlbRingBufferWriteGuard<'_> {
+        let write_lock = self.write_lock.lock();
+        AtomicTlbRingBufferWriteGuard {
+            buf: self,
+            _write_lock: write_lock,
+        }
+    }
+
+    fn try_copy(&self, start_count: Wrapping<usize>, flush_addrs: &mut [u64]) -> bool {
+        let mut index = start_count;
+        for flush_addr in flush_addrs.iter_mut() {
+            *flush_addr = self.buffer[index.0 % FLUSH_GVA_LIST_SIZE].load(Ordering::Relaxed);
+            index += 1;
+        }
+
+        // Check to see whether any additional entries have been added
+        // that would have caused a wraparound. If so, the local list is
+        // incomplete and the copy has failed.
+        if (Wrapping(self.in_progress_count.load(Ordering::Relaxed)) - start_count).0
+            > FLUSH_GVA_LIST_SIZE
+        {
+            return false;
+        }
+        true
+    }
+}
+
+impl AtomicTlbRingBufferWriteGuard<'_> {
+    pub fn push(&self, v: u64) {
+        debug_assert_eq!(
+            self.buf.in_progress_count.load(Ordering::Relaxed),
+            self.buf.gva_list_count.load(Ordering::Relaxed)
+        );
+        // Adding a new item to the buffer must be done in three steps:
+        // 1. Indicate that an entry is about to be added so that any flush
+        //    code executing simultaneously will know that it might lose an
+        //    entry that it is expecting to see.
+        // 2. Add the entry.
+        // 3. Increment the valid entry count so that any flush code executing
+        //    simultaneously will know it is valid.
+        let count = self.buf.in_progress_count.fetch_add(1, Ordering::Relaxed);
+        self.buf.buffer[count % FLUSH_GVA_LIST_SIZE].store(v, Ordering::Relaxed);
+        self.buf.gva_list_count.fetch_add(1, Ordering::Relaxed);
```

Comment:
These relaxed orderings probably all need to be release, I am guessing.

Reply:
How come? We don't have any acquires anywhere. And IIRC the HCL uses relaxed for all this.

Reply:
We need to make sure the buffer is observed to be written before the GVA list count is observed to have incremented, right? Otherwise the reader won't get the accurate value of page to flush.

I assume you also need acquires on the read side, yes.